packages/plugin/vite/src/util/package.ts

```diff
@@ -0,0 +1,120 @@
+  } catch {
+    return false;
+  }
+}
+
+export async function lookupNodeModulesPaths(root: string, paths: string[] = []): Promise<string[]> {
+  if (!root) return paths;
+  if (!path.isAbsolute(root)) return paths;
+
+  const p = path.join(root, 'node_modules');
+
+  if (await isDirectory(p)) {
+    paths = paths.concat(p);
+  }
+  root = path.join(root, '..');
+
+  return isRootPath(root) ? paths : await lookupNodeModulesPaths(root, paths);
+}
+
+export async function readPackageJson(root = process.cwd()): Promise<PackageJsonManifest> {
+  const packageJsonPath = path.join(root, 'package.json');
+  try {
+    const packageJsonStr = await fs.promises.readFile(packageJsonPath, 'utf8');
+    return JSON.parse(packageJsonStr);
+  } catch (error) {
+    console.error(`parse ${packageJsonPath} failed: ${error}`);
+    throw error;
+  }
+}
+
+export async function resolveDependencies(root: string) {
+  const rootDependencies = Object.keys((await readPackageJson(root)).dependencies || {});
+  const resolve = async (prePath: string, dependencies: string[], collected: Map<string, Dependency> = new Map()) =>
+    await Promise.all(
+      dependencies.map(async (name) => {
+        let curPath = prePath,
+          depPath = null,
+          packageJson = null;
+        while (!packageJson && !isRootPath(curPath)) {
+          const allNodeModules = await lookupNodeModulesPaths(curPath);
+
+          for (const nodeModules of allNodeModules) {
+            depPath = path.join(nodeModules, name);
+            if (fs.existsSync(depPath)) break;
+          }
+
+          if (depPath) {
+            try {
+              packageJson = await readPackageJson(depPath);
+            } catch (err) {
+              // lookup node_modules
+              curPath = path.join(curPath, '..');
+              if (curPath.length < root.length) {
+                console.error(`not found 'node_modules' in root path: ${root}`);
+                throw err;
+              }
+            }
+          }
+        }
+
+        if (!depPath || !packageJson) {
+          throw new Error(`find dependencies error in: ${curPath}`);
+        }
+
+        const result: Dependency = {
+          name,
+          path: {
+            src: depPath,
+            dest: path.relative(root, depPath),
+          },
+          dependencies: [],
+        };
+        const shouldResolveDeps = !collected.has(depPath);
+        collected.set(depPath, result);
+        if (shouldResolveDeps) {
+          result.dependencies = await resolve(depPath, Object.keys(packageJson.dependencies || {}), collected);
+        }
+        return result;
+      })
+    );
+  return resolve(root, rootDependencies);
+}
+
+export async function getFlatDependencies(root = process.cwd()) {
+  const depsTree = await resolveDependencies(root);
+  const depsFlat = new Map<string, SourceAndDestination>();
+
+  const flatten = (dep: Dependency) => {
+    depsFlat.set(dep.path.src, dep.path); // dedup
```

Comment:
Question: won't this flattening cause issues when there are multiple major versions of the same package across the dependency tree? For example, if `some-dependency` requires `chalk@^4.0.0` and `some-other-dependency` requires `chalk@^5.0.0`, I think we're only shipping the last version that we happened to visit in this loop, and that version is not necessarily compatible with all packages that depend on it.

Reply:
Answering my own question after running some tests: this code just collects all dependency paths so they can be copied to the dest folder's `node_modules`, I misread the code as if we were using the dependency _names_ as keys, so this is completely fine. My bad! 😅